### PR TITLE
Collectors.py: Pass ignored_globs to the globbing functions

### PR DESCRIPTION
**Pull request by impmihai**
The script originally collected all files it could find and
then check which of them should be ignored.
Now it sends the ignored globs to the iglob function so we
win some execution time.

Closes: #3655

Extreme speed test: http://imgur.com/a/7BUf3
*This is an extreme case scenario, but it is a good
example for big projects, with a lot of object files
that do not need checking. I just copied the coala
source over and over again in the same directory
and run with all files ignored. I really want a
better real world speed test, if anybody can help.

### CODE REVIEW

File: coalib/parsing/Globbing.py [hunk 1]
```diff
@@ -252,14 +252,20 @@ def _absolute_flat_glob(pattern):
...
     Recursively iterates subdirectories of all levels from dirname
 
+    :param ignored_globs: List of globs to ignore when matching files
     :param dirname: Directory name
     :return:        Iterator that yields files and directory from the given dir
                     and all it's (recursive) subdirectories
     """
+    if ignored_globs != None:
+        for ignored_dir in ignored_globs:
+            if dirname.endswith(ignored_dir):
```
Comment (underyx) [inline coalib/parsing/Globbing.py]: I don't think this will work because

```python
ignored_dir == 'env/**'
```

or

```python
ignored_dir == 'vendor/*/node_modules/**'
```

etc.

Comment (impmihai) [inline coalib/parsing/Globbing.py]: @underyx I was asking if the method i am parsing the ignored dirs is ok. I am working already on what you pointed.

Comment (underyx) [inline coalib/parsing/Globbing.py]: Oh, right, sorry, I didn't realize.

I think this method is the best we can do indeed, though I'm not super familiar with this code. It might turn out to be a good idea to do the ignoring elsewhere. If I were implementing this I'd just try to handle the ignoring as soon as possible and run benchmarks to make sure I'm not missing some alternative code path that won't have the ignoring.

File: coalib/parsing/Globbing.py [hunk 2]
```diff
@@ -252,14 +252,20 @@ def _absolute_flat_glob(pattern):
     return
```
Comment (Adrianzatreanu) [inline coalib/parsing/Globbing.py]: commit message: globes-> globs ;p

File: coalib/parsing/Globbing.py [hunk 3]
```diff
@@ -252,14 +252,20 @@ def _absolute_flat_glob(pattern):
...
+def _iter_relative_dirs(dirname, ignored_globs=None):
     """
     Recursively iterates subdirectories of all levels from dirname
 
+    :param ignored_globs: List of globs to ignore when matching files
     :param dirname: Directory name
     :return:        Iterator that yields files and directory from the given dir
                     and all it's (recursive) subdirectories
     """
+    if ignored_globs is not None:
```
Comment (Adrianzatreanu) [inline coalib/parsing/Globbing.py]: wouldnt "if ignored_globs:" be enough?

**Comment by impmihai** [comment]
@underyx Can you run a test with this change like you did in the issue? I don't have a large enough project at the moment to try a benchmark.

File: coalib/parsing/Globbing.py [hunk 4]
```diff
@@ -252,14 +252,20 @@ def _absolute_flat_glob(pattern):
...
+        for ignored_dir in ignored_globs:
+            if fnmatch(dirname, ignored_globs):
```
Comment (userzimmermann) [inline coalib/parsing/Globbing.py]: Should be `if fnmatch(dirname, ignored_dir):`

File: coalib/parsing/Globbing.py (hunk 3 above)
Comment (userzimmermann) [inline coalib/parsing/Globbing.py]: explicit `None` check is just more recommended

File: coalib/parsing/Globbing.py [hunk 5]
```diff
@@ -252,14 +252,20 @@ def _absolute_flat_glob(pattern):
...
+                return
```
Comment (userzimmermann) [inline coalib/parsing/Globbing.py]: Should be `if fnmatch(dirname, ignored_dir):`

File: coalib/parsing/Globbing.py [hunk 6]
```diff
@@ -252,31 +252,37 @@ def _absolute_flat_glob(pattern):
     return
 
 
-def _iter_relative_dirs(dirname):
+def _iter_relative_dirs(dirname, ignored_globs=None):
     """
     Recursively iterates subdirectories of all levels from dirname
 
+    :param ignored_globs: List of globs to ignore when matching files
```
Comment (aptrishu) [inline coalib/parsing/Globbing.py]: maintain the order of param definition according to the argument..
```Cython
:param dirname:          Directory name
:param ignored_globs: List of globs to ignore when matching files
:return:                         Iterator that yields files and directory from the given dir
                                    and all it's (recursive) subdirectories
```
Also maintain the alignment..

Comment (aptrishu) [inline coalib/parsing/Globbing.py]: everywhere..^^

**Comment by sils** [comment]
- you got conflicts
- https://github.com/coala/coala/issues/3655#issuecomment-277850830
- you seem to have a second shortlog right in the middle of your commit message, why? Maybe some squashing, forgot to remove it?

**Comment by sils** [comment]
seems you lack some testing

File: coalib/parsing/Globbing.py [hunk 7]
```diff
@@ -252,40 +252,47 @@ def _absolute_flat_glob(pattern):
...
+    :return:              Iterator that yields files and directory from the
+                          given dir all it's (recursive) subdirectories
     """
     if not dirname:
         dirname = os.curdir
+
+    if ignored_globs:
+        if fnmatch(dirname, ignored_globs):
+            return
```
Comment (RohanVB) [inline coalib/parsing/Globbing.py]: Tests are failing due to coverage by this. 

> 268-269, 267->268

File: coalib/parsing/Globbing.py [hunk 8]
```diff
@@ -252,40 +252,47 @@ def _absolute_flat_glob(pattern):
...
-                    and all it's (recursive) subdirectories
+    :param dirname:       Directory name
+    :param ignored_globs: List of globs to ignore when matching files
+    :return:              Iterator that yields files and directory from the
+                          given dir all it's (recursive) subdirectories
     """
     if not dirname:
         dirname = os.curdir
+
+    if ignored_globs:
```
Comment (Makman2) [inline coalib/parsing/Globbing.py]: what about having an empty tuple as the default value for `ignored_globs`, then you don't need the if^^ `fnmatch` would return immediately because there aren't any globs^^

File: coalib/parsing/Globbing.py [hunk 9]
```diff
@@ -358,49 +368,51 @@ def _iglob(pattern):
...
+    :param ignored_globs: List of globs to ignore when matching files
+    :return:              Iterator that yields all file names that match pattern
     """
     for pat in _iter_alternatives(pattern):
         pat = os.path.expanduser(pat)
         pat = os.path.normcase(pat)
 
         if pat.endswith(os.sep):
-            for name in _iglob(pat):
+            for name in _iglob(pat, ignored_globs):
```
Comment (Makman2) [inline coalib/parsing/Globbing.py]: just for reference, we already supported ignoring globs directly, but we just did not make use of it? :O

**Comment by impmihai** [comment]
Ignored globs were filtered after all files were collected. Initially we were looking for all files and directories in the project and then remove the files and directories that were matching ignore globs.

File: coalib/parsing/Globbing.py (hunk 9 above)
Comment (impmihai) [inline coalib/parsing/Globbing.py]: Ignored globs were filtered after all files were collected. Initially we were looking for all files and directories in the project and then remove the files and directories that were matching ignore globs.

File: coalib/parsing/Globbing.py (hunk 8 above)
Comment (impmihai) [inline coalib/parsing/Globbing.py]: I used it this way to make this parameter optional for functions. This approach means forcing all function calls to have an empty touple added.

File: coalib/parsing/Globbing.py (hunk 7 above)
Comment (impmihai) [inline coalib/parsing/Globbing.py]: Thank you for the help!

File: coalib/parsing/Globbing.py (hunk 8 above)
Comment (Makman2) [inline coalib/parsing/Globbing.py]: why that?
```
def _iter_relative_dirs(dirname, ignored_globs=tuple()):
    ...
```

File: coalib/parsing/Globbing.py (hunk 9 above)
Comment (Makman2) [inline coalib/parsing/Globbing.py]: ah never mind I understand it now :3 Underestimated the `fnmatch` :3

File: coalib/parsing/Globbing.py (hunk 8 above)
Comment (impmihai) [inline coalib/parsing/Globbing.py]: Oh, right, sorry. 
Should i change it like that? It is not that big of a difference.

Comment (Makman2) [inline coalib/parsing/Globbing.py]: yeah please, even though it's just a single line, we avoid it :)

Comment (impmihai) [inline coalib/parsing/Globbing.py]: The code that was already there had ignored_globs initializated with None. Should i change everywhere it's default value then? Because if i don't, it will end up as None sometimes in the _iter_relative_dirs function

Comment (Makman2) [inline coalib/parsing/Globbing.py]: Where exactly?

Comment (impmihai) [inline coalib/parsing/Globbing.py]: The first time it can be initializated with `None` is in `coalib/collecting/Collectors.py' on icollect function

Comment (Makman2) [inline coalib/parsing/Globbing.py]: hm... okay then keep it :+1:

File: coalib/parsing/Globbing.py (hunk 7 above)
Comment (Makman2) [inline coalib/parsing/Globbing.py]: Where does the current file-filtering happen actually?

Comment (impmihai) [inline coalib/parsing/Globbing.py]: Im collectors.py, right after my change in that file

Comment (Makman2) [inline coalib/parsing/Globbing.py]: If possible, I think we should also filter out file names, to make the globbing functions work more consistently (as I respect from `glob` to filter out files too that don't match). Then you can remove the filtering from Collectors.py (which should only deal with collection, not with handling globbing).

Comment (impmihai) [inline coalib/parsing/Globbing.py]: ok, I am working on it, but i think this will take a little time

Comment (impmihai) [inline coalib/parsing/Globbing.py]: @Makman2 looks like it is pretty hard to do that because of cache, so i would have to change lots of tests to make this work the way you asked. That final filter would better stay there

Comment (Makman2) [inline coalib/parsing/Globbing.py]: alright then let's keep it for now :) Be sure to file an issue after merge that we should also filter out files, like someone would expect from this function :+1:

File: coalib/parsing/Globbing.py [hunk 10]
```diff
@@ -336,14 +345,15 @@ def has_wildcard(pattern):
     """
     Checks whether pattern has any wildcards.
 
-    :param pattern: Glob pattern that may contain wildcards
-    :return:        Boolean: Whether or not there are wildcards in pattern
+    :param pattern:       Glob pattern that may contain wildcards
+    :param ignored_globs: List of globs to ignore when matching files
```
Comment (Makman2) [inline coalib/parsing/Globbing.py]: hah too enthusiastic :D
this function does not have an `ignored_globs` param

File: tests/parsing/GlobbingTest.py [hunk 11]
```diff
@@ -222,8 +222,9 @@ class GlobTest(unittest.TestCase):
     def setUp(self):
         self.maxDiff = None
 
-    def _test_glob(self, pattern, file_list):
-        results = sorted([os.path.normcase(g) for g in glob(pattern)])
+    def _test_glob(self, pattern, file_list, ignored_globs=None):
```
Comment (Makman2) [inline tests/parsing/GlobbingTest.py]: huh why do you add an `ignored_globs` param? :O

File: coalib/parsing/Globbing.py (hunk 10 above)
Comment (impmihai) [inline coalib/parsing/Globbing.py]: Whoops:D

File: tests/parsing/GlobbingTest.py (hunk 11 above)
Comment (Makman2) [inline tests/parsing/GlobbingTest.py]: aaaah nvm... why the heck is there just an underscore before the function name, I thought it was a test... not your fault :3

**Comment by impmihai** [comment]
Anything else to add? 😄

File: coalib/collecting/Collectors.py [hunk 12]
```diff
@@ -52,7 +52,7 @@ def icollect(file_paths, ignored_globs=None, match_cache={}):
 
     for file_path in file_paths:
         if file_path not in match_cache:
-            match_cache[file_path] = list(iglob(file_path))
+            match_cache[file_path] = list(iglob(file_path, ignored_globs))
 
         for match in match_cache[file_path]:
             if not ignored_globs or not fnmatch(match, ignored_globs):
```
Comment (sils) [inline coalib/collecting/Collectors.py]: is this loop needed then?

File: coalib/parsing/Globbing.py [hunk 13]
```diff
@@ -299,33 +306,35 @@ def relative_wildcard_glob(dirname, pattern):
     return result
 
 
-def relative_flat_glob(dirname, basename):
+def relative_flat_glob(dirname, basename, ignored_globs=None):
```
Comment (sils) [inline coalib/parsing/Globbing.py]: this function doesn't need it!

File: coalib/parsing/Globbing.py [hunk 14]
```diff
@@ -358,49 +367,51 @@ def _iglob(pattern):
...
+def iglob(pattern, ignored_globs=None):
     """
     Iterates all filesystem paths that get matched by the glob pattern.
     Syntax is equal to that of fnmatch.
 
-    :param pattern: Glob pattern with wildcards
-    :return:        Iterator that yields all file names that match pattern
+    :param pattern:       Glob pattern with wildcards
+    :param ignored_globs: List of globs to ignore when matching files
+    :return:              Iterator that yields all file names that match pattern
```
Comment (sils) [inline coalib/parsing/Globbing.py]: you should elaborate on the ignore here - you still return some ignored files, right?You just don't traverse into ignored dirs.

Comment (Makman2) [inline coalib/parsing/Globbing.py]: I already noticed this, and I think we can make a new issue for that. Let's improve speed for now, and increase consistency in a separate PR, so we get something merged.

File: coalib/parsing/Globbing.py (hunk 13 above)
Comment (impmihai) [inline coalib/parsing/Globbing.py]: yes, but in globbing.py in _iglob function, there is a variable names relative_glob_function which can get this function assigned. And that relative_glob_function is called with 2 parameters.

Comment (Makman2) [inline coalib/parsing/Globbing.py]: don't understand that. If you don't need it, then you don't need, am I wrong? :D

Comment (impmihai) [inline coalib/parsing/Globbing.py]: `if basename == '**':
    relative_glob_function = relative_recursive_glob
elif has_wildcard(basename):
    relative_glob_function = relative_wildcard_glob
else:
    relative_glob_function = relative_flat_glob
 
if not dirname:
    for file in relative_glob_function(dirname, basename, ignored_globs):
        yield fille`

This is how the function looks like. And in case the if exists on 3rd path, the `relative_glob_function` is equal to `relative_flat_glob`. After this, the `relative_glob_function` is called with 2 parameters. So `relative_flat_glob` gets called with 2 parameters. If it expects only one, then it starts complaining.

File: coalib/parsing/Globbing.py [hunk 15]
```diff
@@ -336,14 +345,14 @@ def has_wildcard(pattern):
     """
     Checks whether pattern has any wildcards.
 
-    :param pattern: Glob pattern that may contain wildcards
-    :return:        Boolean: Whether or not there are wildcards in pattern
+    :param pattern:       Glob pattern that may contain wildcards
```
Comment (Asnelchristian) [inline coalib/parsing/Globbing.py]: uhm... this should go in a different commit

File: coalib/parsing/Globbing.py [hunk 16]
```diff
@@ -336,14 +345,14 @@ def has_wildcard(pattern):
...
+    :return:              Boolean: Whether or not there are wildcards in pattern
```
Comment (Asnelchristian) [inline coalib/parsing/Globbing.py]: this change is not actually needed

**Comment by Asnelchristian** [comment]
this is okay for me :+1:

**Comment by Asnelchristian** [comment]
needs rebase

**Comment by Makman2** [comment]
Closing due to inactivity.
